### PR TITLE
cycript 0.9.592 (new formula)

### PR DESCRIPTION
**Pull request by conradev**

### CODE REVIEW

File: Library/Formula/cycript.rb [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  homepage "http://www.cycript.org"
+  url "git://git.saurik.com/cycript.git", :tag => "v0.9.592", :revision => "97bec96b43b66ab78af95d2b6c6c24f0d0a8006a"
+
+  head "git://git.saurik.com/cycript.git"
+
+  depends_on "bison" => :build
+  depends_on "android-sdk" => :build
+  depends_on "readline"
+
+  patch :DATA
```
Comment (DomT4) [inline Library/Formula/cycript.rb]: Can you link to an upstream bug submission/etc and explain briefly what this is fixing here? Thanks!

File: Library/Formula/cycript.rb [hunk 2]
```diff
@@ -0,0 +1,93 @@
+class Cycript < Formula
+  desc "Modify apps using a hybrid of Objective-C++ and JavaScript syntax."
+  homepage "http://www.cycript.org"
+  url "git://git.saurik.com/cycript.git", :tag => "v0.9.592", :revision => "97bec96b43b66ab78af95d2b6c6c24f0d0a8006a"
```
Comment (DomT4) [inline Library/Formula/cycript.rb]: Please break these onto their own lines:
```ruby
url "git://git.saurik.com/cycript.git",
    :tag => "v0.9.592",
    :revision => "97bec96b43b66ab78af95d2b6c6c24f0d0a8006a"
```

File: Library/Formula/cycript.rb [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+  depends_on "bison" => :build
+  depends_on "android-sdk" => :build
+  depends_on "readline"
+
+  patch :DATA
+
+  def install
+    ENV.append "CPPFLAGS", "-Wno-inconsistent-missing-override"
+    ENV.append "CPPFLAGS", "-I#{HOMEBREW_PREFIX}/opt/readline/include"
```
Comment (DomT4) [inline Library/Formula/cycript.rb]: These two lines (The `readline` ones) shouldn't be necessary. Should happen automatically.

File: Library/Formula/cycript.rb [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    ENV.append "LDFLAGS", "-L#{HOMEBREW_PREFIX}/opt/readline/lib"
+
+    system "./configure",
+           "--prefix=#{prefix}",
+           "CPPFLAGS=#{ENV.cppflags}",
+           "LDFLAGS=#{ENV.ldflags}",
+           "BISON=#{HOMEBREW_PREFIX}/opt/bison/bin/bison"
```
Comment (DomT4) [inline Library/Formula/cycript.rb]: Same on this line. Should be picking up this version automatically?

File: Library/Formula/cycript.rb [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+    system "make", "install"
+  end
+
+  test do
+    assert_equal "7", `echo 'print([@[@2,@3].map(a => a + 1) valueForKeyPath:@"@sum.self"])' | #{bin}/cycript -`.strip
```
Comment (bfontaine) [inline Library/Formula/cycript.rb]: Use `pipe_output("#{bin}/cycript -", "...")` instead of `` `echo '...' | #{bin}/cycript -`.

File: Library/Formula/cycript.rb (hunk 1 above)
Comment (conradev) [inline Library/Formula/cycript.rb]: `cycript` requires a few files to operate: `libcycript.db`, `libcycript.jar` and `libcycript.dylib`

Normally, it places them in `#{prefix}/lib`, but `brew audit` does not like that at all (`.db` is not a library, and `.jar` files are not allowed in `lib`). I patched the code that looks for these files and the code that installs them to place them in `#{prefix}/share` instead.

It's really a brew-specific change, and I haven't discussed the proper solution with @saurik yet. Personally, I think they should be able to go in `#{prefix}/lib`.

**Comment by conradev** [comment]
Updated diff to account for all changes!

File: Library/Formula/cycript.rb (hunk 1 above)
Comment (bfontaine) [inline Library/Formula/cycript.rb]: A good solution would be to be able to tell the build process where these files are installed so any path would work.

Comment (conradev) [inline Library/Formula/cycript.rb]: @bfontaine a patch would still be required to not install them to `#{prefix}/lib` and to install them somewhere else

But is the ideal patch one where these options can be passed to configure?

File: Library/Formula/cycript.rb [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  head "git://git.saurik.com/cycript.git"
+
+  depends_on "bison" => :build
+  depends_on "android-sdk" => :build
+  depends_on "readline"
+
+  patch :DATA
+
+  def install
+    ENV.append "CPPFLAGS", "-Wno-inconsistent-missing-override"
```
Comment (bfontaine) [inline Library/Formula/cycript.rb]: Why is this necessary for the build?

File: Library/Formula/cycript.rb [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+diff --git a/Execute.cpp b/Execute.cpp
+index 18bdce6..cb850f1 100644
+--- a/Execute.cpp
++++ b/Execute.cpp
+@@ -2153,7 +2153,7 @@ void CYInitializeDynamic() {
+     else return;
+ 
+     CYPool pool;
+-    const char *db(pool.strcat(CYPoolLibraryPath(pool), "/libcycript.db", NULL));
++    const char *db(pool.strcat(CYPoolLibraryPath(pool), "/../share/libcycript.db", NULL));
```
Comment (bfontaine) [inline Library/Formula/cycript.rb]: Is this something that’ll change on version bumps? I mean if you update `cycript`, do you expect this file to be overrided by the new version or should that be something we store outside of the prefix?

File: Library/Formula/cycript.rb (hunk 1 above)
Comment (bfontaine) [inline Library/Formula/cycript.rb]: We usually prefere `inreplace` calls instead of a patch for Homebrew-specific stuff, so that gives:
```rb
# install support files in share instead of lib
inreplace "Execute.cpp", '"/libcycript.db"', '"/../share/libcycript.db"'
inreplace "Execute.cpp", '"/libcycript.cy"', '"/../share/libcycript.cy"'
inreplace "Java/Execute.cpp", '"/libcycript.jar"', '"/../share/libcycript.jar"'
inreplace "Makefile.am", "datdir = $(libdir)", "datdir = $(datadir)"
inreplace "Makefile.in", "datdir = $(libdir)", "datdir = $(datadir)"
```

**Comment by saurik** [comment]
I am the developer of Cycript. (FWIW, I purposely have never attempted to submit my project to Homebrew, though you are clearly welcome to try to make this work yourself. Understand, though, that I have never tested any of Cycript against any of the libraries or tools in Homebrew, and have some philosophical reasons for why I will never be able to do so myself: I refuse to install Homebrew on any computer I own. Cycript building on Mac OS X has been developed exclusively using Fink.)

Regardless, I feel the need to chime in at least momentarily here, as this patch is a bad idea, not because it is a workaround for some deficiency in my project that would somehow be better fixed with an upstream patch, but because it is acting as a workaround for what is apparently (assuming I am reading this thread correctly, that there is some tool that is refusing this installation) some broken understanding in Homebrew's tooling about the purpose of these folders :/. libcycript.db is built for the target architecture(s) and thereby cannot be shared even among the same operating system: this means it *must not* go in "share" :(. You can argue that this file could (or maybe even "should"? though I do not currently believe this to be true) go in a subfolder of lib, but it absolutely must go under the lib hierarchy.

I can see a separate argument that libcycript.jar's current implementation could probably be engineered in a way so it could be stored in share, but I make no such guarantee that this will remain true the future (and in fact intend this not to be the case going forward, so I can drop the dependency on Android's dx), and I further believe it is technically incorrect as it stands right now (as I don't think I override the javac target version).

I can also see an argument that it would be allowed to put libcycript.cy (which I see actually isn't being correctly installed by configure/make... that actually is a bug in Cycript...) under share, but that would at this point be "unorthodox" (per other scripting languages, including as one example Python) to not even have a symlink from lib. (That said, I would be much more sympathetic here, in contrast to libcycript.db, to an argument of moving libcycript.cy either under a subfolder in lib. Though at that point I would probably instead move it into libcycript.db.)

File: Library/Formula/cycript.rb (hunk 6 above)
Comment (saurik) [inline Library/Formula/cycript.rb]: Clang seems to have added some kind of new warning. I guess I should really disable -Wall if you are not in maintainer mode, as otherwise every time someone uses a compiler different than one of the ones I have directly tested with, the build is subject to randomly fail. (I am used to just telling people what compilers I am compatible with, but that is no longer a viable situation for Cycript, and is no longer necessary since I moved all the insane parts out to a different in-coupled project.) I'll fix that at some point.

Comment (UniqMartin) [inline Library/Formula/cycript.rb]: > I guess I should really disable -Werror if you are not in maintainer mode, […]

Yes, you should! `-Werror` is really only a developer tool and should never be used when distributing software. It is a good goal to eliminate all warnings during development and `-Werror` helps with this, but you never know what warnings there will be in a future or untested compiler. (Or maybe caused by one of the libraries you use.)

**Comment by DomT4** [comment]
Hey Jay. It's nice to see your long comments persist across multiple social platforms! :wink:

I won't get into the philosophical differences of package management; what people want to use and why is entirely up to them and we'll likely bikeshed this PR to death if we get into it.

> some broken understanding in Homebrew's tooling about the purpose of these folders

We have a check to ensure `.jar` files aren't dumped in the globally-available prefix because they _tend_ (although not particularly in this case) to share rather common names and we've discovered the potential for conflict on that over time and now consequently attempt to work around it with nesting.

Rather than patching, what we tend to do in situations more or less identical to this is install all the files in the subdirectory `libexec`, which isn't symlinked to the globally-available prefix, so the files live where they expect to live in terms of the `.jar` existing in a `lib` directory and then create an exec script from the executable in `libexec/bin` to `prefix/bin` which _is_ symlinked to the globally-available prefix.

I suspect, glancing this over, that system would work here. OP should be able to test this locally by changing:
```ruby
"--prefix=#{prefix}",
```
to:
```ruby
"--prefix=#{libexec}",
```
And then adding:
```ruby
bin.write_exec_script Dir["#{libexec}/bin/*"]
```
After the `make install` stage.

**Comment by conradev** [comment]
I pushed an amended version of the formula that does as @DomT4 suggests, and it works great.

I also added more tests which check for the presence of `libcycript.cy` and `libcycript.db` programmatically. These tests currently fail because `libcycript.cy` is not included in the install phase.

I will write a patch and send it to @saurik soon.

**Comment by apjanke** [comment]
Ping on the patch?

Closing because legacy-homebrew is deprecated. If you find time to work on this, please resubmit to the new homebrew/core repo.
